File icon

y-lohse opened the pull request:
We want to display certain files from the cozy on the homepage (specifically some .url files). However we'd like the ability to display those files with a custom icon, instead of a generic icon or the site's favicon. The plan is to insert the icon SVG in the couch document (being icons, the svgs are quite small).

I suggest we name the new entry `icon`… because it's an icon and I don't see a reason to do something more complicated than that.

I don't think `icon` should be inside `cozyMetadata` since the information is not about the couch document. Since the icon is not really derived from the file itself, I wouldn't put it in `metadata` either, so I suggest adding it to the file data, along things such as `trashed` and `size` for example.

nono commented on docs/io.cozy.files.md:
```diff
@@ -34,6 +34,7 @@ The attributes of a file are:
 - `executable`: {bool} true is the file has the executable bit on UNIX (`chmod +x`)
 - `class`: {string} a class in the list: `['image', 'document', 'audio', 'video', 'text', 'binary']`
 - `mime`: {string} the full mime-type
+- `icon`: {string} : Optionnal, a small svg file or the URL of an image to use as icon for this file
```
s/Optionnal/optional/

y-lohse replied:
```suggestion
- `icon`: {string} : Optional, a small SVG file or the URL of an image to use as icon for this file
```

Crash-- commented:
So the rule is : 

Check if `icon` (I agreed with @nono I prefer to have it in metadata.icon) is defined if yes use it, else fetch the `links` ?

y-lohse commented:
> Check if icon (I agreed with @nono I prefer to have it in metadata.icon) is defined if yes use it, else fetch the links ?

The links only exist for photo files, so yes but not for all file types. Generally speaking, we haven't decided that the icon should be used *all* the time, especially in Drive. But yeah the rule would probably be `icon` > `links` > `cozy-ui file type icon`

I think `metadata` is reasonable too, but the only good argument I see is because it would be an optional field. But `path` is also an optional field, so… I still feel like `icon` is closer to `size` than `width` for example..?

y-lohse commented:
I realize now that there is also a security implication: the SVG would be injected into the DOM straight from the database. This could be a big problem if the SVG contains some code.

i don't really trust that we can parse out code from a SVG string. It wouldn't be too hard to forge a document with a malicious `icon` field. And through sharings it could land on other cozy's. the more I think about it the less i'm comfortable with that.

If the icon field is a link to an image, I think those risks are eliminated. But can we load an arbitrary image or will the CSPs block us @nono? If we get blocked, where would be a good place to host those icons?

nono commented:
I don't think it is possible to execute JS via `<img src="data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDo etc" />`. And even if it is possible, the CSP will block it on cozy instances.

> But can we load an arbitrary image or will the CSPs block us @nono?

The CSP will let you load an image only from an authorized host (ie on the cozy its-self).

> If we get blocked, where would be a good place to host those icons?

Maybe via the dynamic assets of the stack (https://docs.cozy.io/en/cozy-stack/config/#assets). But if the file is shared to a cozy instance on a different context, it won't work.

edas commented:
For SVG that we can't review individually ourselves:

A should never embed them directly inside the DOM of our pages. It would bea security risk if wee display an svg from user A inside the cozy of user B, but also if the current user got himself its SVG from a third party.

We could link the SVG inside a <img> tag, but at the strict condition that no-one outside the current user may know or predict the url of the SVG file itself.

The scenario to avoid : Alices manage to have Bob putting a malicious SVG for one of its cozy-files. If Alice is able to predict the URL of the SVG file itself, she could make Bob click on a link to this SVG, and make Bob display the SVG in full page, executing any Javascript in it.

nono commented:
> We should never embed them directly inside the DOM of our pages. It would bea security risk if wee display an svg from user A inside the cozy of user B, but also if the current user got himself its SVG from a third party.

Do you mean that `<img src="data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDo etc" />` where the data-URI is controlled by user A is a security risk? If yes, I would like to have more details.

edas commented:
No, by embedding directly I was speaking of on svg tag directly inside the DOM (by opposition with linking the svg inside an img tag)

y-lohse commented:
So am I correct in saying that we could store the data-uri representation of the SVG in couch and use that inside and `img` tag and it would be acceptable? Because this solution is a lot easier than hosting files somewhere and linking to them in the document.

edas commented:
> So am I correct in saying that we could store the data-uri representation of the SVG in couch and use that inside and `img` tag and it would be acceptable? Because this solution is a lot easier than hosting files somewhere and linking to them in the document.

Yep, I think so. 

Solutions I see: 

* Use the stack JSON API to get a base64 data, and create a &lt;img src="data:xxx"  /&gt;.  Its OK from a security point of view, but may suffer from performance problems if we reuse too much times the  same icon on a page, or if the icon is really heavy.

* Similarly, get the binary content into a blob from the stack and create a &lt;img src="blob:xxx"  /&gt; with createObjectURL. It should be a little better if we reuse the same icon multiple times or if the icon is really heavy.

* Use a system like the thumbnails we already have, request a public temporary link to the icon, and display a &lt;img src="https://xxx" /&gt;. We could even request a thumbnail in png here.

* Host icons in a special domain like xxxx-thirdparty.mycozy.cloud with no cookie and link them in &lt;img src="https://xxx-thirdparty.mycozy.cloud/xxxx" /&gt;

* Serve public icons with a permanent URI from the stack, but make the URL absolutly not predicatable by anyone.

y-lohse commented:
After a quick chat, our 3 mains options are:

1. Store the svg in couch, like `icon: "<svg>etc…"`, and display it in an image tag with a plain text data-uri (good size savings and perfs, but risk of rogue quotes).
2. Same thing but encoding the SVG to base64 and using that as data-uri. Or even saving the icon straight as a base64 string in couch.
3. Same thing, but converting the SVG to a BLOB first and using `createObjectURL `. This is great if the icon is used multiple times, but in this case each icon will probably only be displayed once per page.

I'll do some tests but I suspect the second solution is the right compromise for this situation.

y-lohse commented:
I updated the proposal:

- `icon` is now inside `metadata`, mostly because the `/file` API does not seem to return it if it's in the "normal" section of the document
- It's now an SVG and nothing else
- Added an example

On the front-end, we convert it to base64 and use it as a data URI 👍
